src/libraries/System.Text.Json/src/System/Text/Json/Serialization/PreserveReferenceResolver.cs

```diff
@@ -0,0 +1,73 @@
+﻿// Licensed to the.NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+
+namespace System.Text.Json.Serialization
+{
+    /// <summary>
+    /// The default ReferenceResolver implementation to handle duplicate object references.
+    /// </summary>
+    internal sealed class PreserveReferenceResolver : ReferenceResolver
+    {
+        private uint _referenceCount;
+        private readonly Dictionary<string, object>? _referenceIdToObjectMap;
```

Comment:
I assume `ConcurrentDictionary` is not used is because we create a new instance for every (de)serialization call -- correct? i.e. we don't have to worry about multiple threads

Reply:
That's correct. We call `CreateConverter` that returns a new instance every time we (de)serialize unless someone provides its own `ReferenceHandler` implementation that does otherwise.